cmd/vsphere-xcopy-volume-populator/vsphere-xcopy-volume-populator.go

```diff
@@ -313,3 +350,169 @@ func setupTracing() (*prometheus.CounterVec, error) {
 	return progressCounter, nil
 
 }
+
+// getOrGenerateProviderSSHKeys retrieves SSH keys from provider-specific secrets, environment variables, or generates them if not present
+func getOrGenerateProviderSSHKeys(providerHostname string) ([]byte, []byte, error) {
+	if clientSet != nil {
+		privateKeyBytes, publicKeyBytes, err := getSSHKeysFromProviderSecret(clientSet, providerHostname)
+		if err == nil {
+			return privateKeyBytes, publicKeyBytes, nil
+		}
+		klog.V(2).Infof("SSH keys not found in secret for provider %s: %v", providerHostname, err)
+	}
+
+	sshPrivateKeyEnv := os.Getenv("SSH_PRIVATE_KEY_" + providerHostname)
+	sshPublicKeyEnv := os.Getenv("SSH_PUBLIC_KEY_" + providerHostname)
+
+	if sshPrivateKeyEnv != "" && sshPublicKeyEnv != "" {
+		privateKeyBytes, err := base64.StdEncoding.DecodeString(sshPrivateKeyEnv)
+		if err != nil {
+			return nil, nil, fmt.Errorf("failed to decode SSH private key for provider %s: %w", providerHostname, err)
+		}
+
+		publicKeyBytes, err := base64.StdEncoding.DecodeString(sshPublicKeyEnv)
+		if err != nil {
+			return nil, nil, fmt.Errorf("failed to decode SSH public key for provider %s: %w", providerHostname, err)
+		}
+
+		return privateKeyBytes, publicKeyBytes, nil
+	}
+
+	klog.Infof("Generating new SSH key pair for provider %s", providerHostname)
+	privateKey, err := rsa.GenerateKey(rand.Reader, 2048)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to generate RSA key for provider %s: %w", providerHostname, err)
+	}
+
+	privateKeyPEM := &pem.Block{
+		Type:  "RSA PRIVATE KEY",
+		Bytes: x509.MarshalPKCS1PrivateKey(privateKey),
+	}
+	privateKeyBytes := pem.EncodeToMemory(privateKeyPEM)
+
+	publicKey, err := ssh.NewPublicKey(&privateKey.PublicKey)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to create SSH public key for provider %s: %w", providerHostname, err)
+	}
+
+	publicKeyBytes := ssh.MarshalAuthorizedKey(publicKey)
+
+	if err := storeSSHKeysInProviderSecret(clientSet, providerHostname, privateKeyBytes, publicKeyBytes); err != nil {
```

Comment:
reminder that this is a race condition and we need to switch to a 'push' model where the key is generated by the provider controller and it pushes it, using the mergeSecrets into the popualtor secret.